### ISSUE TITLE
fiberassign._internal.Hardware failure

### ISSUE THREAD

**Issue by sbailey** [desihub/fiberassign#244]
Using fiberassign master as installed under desiconda/20190804-1.3.0-spec, I'm getting an error instantiating the Hardware model.  This clone is updated nightly with "git pull" and "python setup.py build_ext --inplace" but it does *not* do a "python setup.py clean".   This has probably left it in a incompatible binary state after the NERSC "upgrade" this past week, though also see issue #243 for problems when starting from a fresh build.
```
$> source /project/projectdirs/desi/software/desi_environment.sh master
$> cd /global/cscratch1/sd/sjbailey/desi/test/fiberassign_mock
$> fiberassign --overwrite --mtl mtl-dark.fits --sky sky.fits --footprint ./tiles.fits --outdir ./output/

INFO: Loaded focalplane for time stamp 2019-12-08 05:51:12.287096
Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20190804-1.3.0-spec/code/fiberassign/master/bin/fiberassign", line 151, in <module>
    run_assign_full(assign_args)
  File "/global/common/software/desi/cori/desiconda/20190804-1.3.0-spec/code/fiberassign/master/py/fiberassign/scripts/assign.py", line 329, in run_assign_full
    hw, tiles, tgs = run_assign_init(args)
  File "/global/common/software/desi/cori/desiconda/20190804-1.3.0-spec/code/fiberassign/master/py/fiberassign/scripts/assign.py", line 249, in run_assign_init
    hw = load_hardware(rundate=args.rundate)
  File "/global/common/software/desi/cori/desiconda/20190804-1.3.0-spec/code/fiberassign/master/py/fiberassign/hardware.py", line 150, in load_hardware
    [positioners[x]["petal"] for x in locations])
TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. fiberassign._internal.Hardware(location: List[int], petal: List[int], device: List[int], slitblock: List[int], blockfiber: List[int], fiber: List[int], device_type: List[str], x: List[float], y: List[float], status: List[int], theta_offset: List[float], theta_min: List[float], theta_max: List[float], theta_arm: List[float], phi_offset: List[float], phi_min: List[float], phi_max: List[float], phi_arm: List[float], excl_theta: List[fiberassign._internal.Shape], excl_phi: List[fiberassign._internal.Shape], excl_gfa: List[fiberassign._internal.Shape], excl_petal: List[fiberassign._internal.Shape])

Invoked with: '2019-09-16T00:00:00', array([7000, 7001, 7002, ..., 6524, 6525, 6526], dtype=int32), <Column name='PETAL' dtype='int32' description='Petal location [0-9]' length=5020>
  7
  7
  7
  7
  7
  7
  7
  7
  7
...
  6
  6
  6
  6
  6
  6
  6
  6, <Column name='DEVICE' dtype='int32' description='Device location on the petal' length=5020>
  0
  1
  2
  3
  4
  5
  6
  7
  8
...
519
520
521
522
523
524
525
526, <Column name='SLITBLOCK' dtype='int32' description='The slit block where this fiber goes' length=5020>
 18
 18
 18
 18
 18
 19
 19
 18
 19
...
 18
 19
 18
 18
  0
  1
  0
  1, <Column name='BLOCKFIBER' dtype='int32' description='The fiber index within the slit block' length=5020>
 20
  4
 21
 13
 17
  0
 23
 22
 12
...
  1
 13
 18
 10
  0
  8
  6
  9, <Column name='FIBER' dtype='int32' description='PETAL * 500 + SLITBLOCK * 25 + BLOCKFIBER' length=5020>
3970
3954
3971
3963
3967
3975
3998
3972
3987
 ...
3451
3488
3468
3460
3000
3033
3006
3034, array([b'POS', b'POS', b'POS', ..., b'POS', b'POS', b'POS'], dtype='|S8'), <Column name='OFFSET_X' dtype='float64' unit='mm' description='X location of positioner center' length=5020>
 -25.81851546044606
 -34.24598804414511
 -35.35336376572603
 -42.67337200252829
 -43.78214641435156
-51.100673731821566
-52.209867804888624
  -53.3373030713626
 -59.52698852979352
                ...
 -243.5292243353441
-250.48870843723313
-257.45934393359687
 -264.4170748433027
 -271.3789356442772
 -278.3472949110281
 -285.3424481850115
-292.34684692341966, <Column name='OFFSET_Y' dtype='float64' unit='mm' description='Y location of positioner center' length=5020>
12.328919779295408
18.451823418204235
 8.067607832475183
24.574660194916063
14.188635165735029
 30.69742981228919
 20.31094455877104
 9.895378215863193
 36.81947372960929
               ...
323.98875963098203
   316.20680479154
 308.4451873655586
 300.6607298001689
 292.8942478157465
285.13133329927103
   277.39196964022
 269.6557351268304, array([0, 0, 0, ..., 0, 0, 0], dtype=uint32), array([-12.7988962 ,  61.9661707 , 153.95265867, ...,  22.18094305,
       -53.94005505, -20.31861998]), array([-196.2914521 , -197.48776528, -196.09305774, ..., -196.07052424,
       -197.48472263, -196.83413549]), array([196.2914521 , 197.48776528, 196.09305774, ..., 196.07052424,
       197.48472263, 196.83413549]), array([3.04995124, 2.89737883, 2.98911091, ..., 2.98153627, 3.01080228,
       3.11710945]), array([-7.33299804, -3.62524903, -9.29825285, ..., -2.0841581 ,
       -2.61267405, -6.98822937]), array([-6.51023712, -3.7366336 , -9.90478481, ..., -4.66532844,
       -2.97917055,  1.48039874]), array([185., 185., 185., ..., 185., 185., 185.]), array([3.09065266, 3.10304677, 3.0094556 , ..., 3.00606651, 3.04019282,
       3.02145895]), [<fiberassign.Shape at (0, 0) has 0 circles and 1 segment collections:
  segment points:
    (0.814, -3.236)
    (2.083, -2.707)
    (2.613, -2.665)
    (4.154, -2.759)
    (4.695, -1.68)
    (5.094, -0.54)
    (5.094, 0.54)
    (4.831, 1.291)
    (4.235, 1.933)
    (3.432, 2.283)
    (2.28, 2.283)
    (-1.902, -0.935)
    (-2.007, -2.665)
    (-1.139, -3.137)
    (-0.17, -3.332)
    (0.814, -3.236)
>, <fiberassign.Shape at (0, 0) has 0 circles and 1 segment collections:
  segment points:
    (0.814, -3.236)
    (2.083, -2.707)
    (2.613, -2.665)
    (4.154, -2.759)
    (4.695, -1.68)
    (5.094, -0.54)
    (5.094, 0.54)
    (4.831, 1.291)
    (4.235, 1.933)
    (3.432, 2.283)
    (2.28, 2.283)
    (-1.902, -0.935)
    (-2.007, -2.665)
    (-1.139, -3.137)
    (-0.17, -3.332)
    (0.814, -3.236)

[etc for thousands of lines]
```

**Comment by tskisner** [comment]
Ok, will investigate this.  For reference in my testing I always do:
```
python setup.py clean
python setup.py install --prefix <blah>
python setup.py test
```
Note that we do have a custom `clean` command:

https://github.com/desihub/fiberassign/blob/master/setup.py#L117

But this does not explicity remove the the "in place" compiled extensions.  I can change that of course.

**Comment by tskisner** [comment]
This looks like an incompatibility between the compiled extensions and the other python code.  This commit about 4 days ago:

https://github.com/desihub/fiberassign/commit/2c482762e8050a75764e1436f59c187dfd33c98c

Changed the C++ interface to the hardware class and the corresponding python code which uses it.  Your error indicates that you have the new python code and a stale compiled extension.  I'll leave this open until I modify the `clean` command to remove any in-place extensions.  However, we should change any automated scripts and documentation to explicitly run the `clean` command before building.

**Comment by sbailey** [comment]
Thanks.  I confirm that the fix for #243 allowed me to to a clean rebuild of the pre-installed fiberassign/master and that also fixed this problem.  Closing.
